### DISCUSSION TITLE
Managing `depends` lists?

### DISCUSSION THREAD

**Discussion by polarathene** [goreleaser/nfpm#636] [Q&A] [answered]
**TL;DR:** When `overrides` is not viable (_same package format, different arch, different packages_), how do you manage multiple package differences? (_different packages, different number of packages_)

---

- A package of the same format can have different package names to depend on depending on distro / release installing the package.
- The nFPM docs only show the YAML syntax for individual list items prefixed with `-`, but the `[ ... ]` syntax also works.
- When package lists are symmetrical you can use ENV more easily to provide appropriate names, but bit problematic when asymmetrical?

ENV:
```bash
# Fedora RPM depends (attempted to provide YAML array list as full value string, rejected):
#REQUIRED_PACKAGES="[ragel, openssl, glib2, sqlite, file-libs, libicu, pcre, zlib, libsodium, luajit, hyperscan]"
# openSUSE RPM depends (attempted to provide YAML array list as items only and wrap ENV in `nfpm.yaml`, rejected):
REQUIRED_PACKAGES="ragel, openssl, glib2, sqlite, libmagic1, libicu, pcre, zlib, libsodium, luajit"
# Example of arch dependent packages:
# x86_64:
REQUIRED_PACKAGES="${REQUIRED_PACKAGES}, hyperscan-devel, jemalloc"
# aarch64:
REQUIRED_PACKAGES="${REQUIRED_PACKAGES}, vectorscan-devel"
```

`nfpm.yaml` (neither works):
```yaml
#depends: ${REQUIRED_PACKAGES}
depends: [ ${REQUIRED_PACKAGES} ]
```

What is the right way to approach this? Do you stitch several YAML files together to provide config, or commit several different config copies with common content repeated?

- AMD64 vs ARM64 differences (_`overrides` feature doesn't seem like it supports this_)
- Differences for same package format (_`overrides` feature doesn't seem like it supports this_)
- Variants (_alternative builds for the same platform + package, but compiled with different build flags_)

`cue` or `ytt` seem like they might be a bit overkill for this, but the ENV expansion also doesn't seem suitable? Am I better to provide overrides with external context instead? Such as `sed` to replace the `depends:` line with a YAML list value?

I haven't tried YAML anchors + aliases yet, perhaps I could store the values in `nfpm.yaml` that way and use an ENV to switch :thinking:

**Comment by polarathene** [comment, marked as answer]
- Empty items (**works**): It seems I can have empty list items, but placing ENV to provide values there may need a generic name if it could represent various packages..
- Anchors/Aliases (**doesn't work**): `nfpm.Config` seems to get validated, any foreign keys defined (_even if using the convention of `x-` prefix like Docker offers for extensions_) throws an error.
- [`yq`](https://github.com/mikefarah/yq) could be used to modify YAML via shell scripts in CI.
- [`envsubst`](https://github.com/a8m/envsubst) seems to be the advised alternative to replace file content ENV vars before `nfpm` processes the config. This was the simplest approach for bulk replacing the `depends` for me :+1:
